### ISSUE TITLE
[rl] No AdamH optimizer: Snowball RL would train an AdamH-trained model with AdamW

### ISSUE THREAD

**Issue by penfever** [marin-community/MarinSkyRL#231]
Snowball/Grug was SFT'd with **AdamH**, which exists only in JAX — `levanter.optim.adamh`, with a local iteration copy at `marin/experiments/grug/moe/adamh.py`. MarinSkyRL has no equivalent. Its optimizer surface is `torch.optim` subclasses plus a special-cased `'Muon'` (`skyrl-train/skyrl_train/distributed/fsdp_strategy.py:499`), and the shipped default is `optimizer: "AdamW"` in `ppo_base_config.yaml`.

So the planned non-agentic RL run on Snowball (marin-community/marin#7786) will continue training, with AdamW, a model that was trained with AdamH.

## Why this may not be a neutral substitution

`marin/experiments/grug/moe/model.py:387` defines a `GatedNorm` module whose docstring is explicit:

> Learnable per-dimension gating. Compensates for AdamH's bounded activation norms.
> See https://arxiv.org/abs/2601.22966v1

The architecture carries a component whose stated purpose is to compensate for a property of the optimizer that trained it. AdamH preserves parameter norm by construction — its update computes `p_norm` and `u_norm` per parameter, steps along the normalized update scaled by `p_norm`, then renormalizes the result back to `p_norm`:

```python
def _scale_invariant_2d(p, u):
    p_norm = jnp.linalg.norm(p)
    u_norm = jnp.linalg.norm(u)
    new_p = p - learning_rate * u * p_norm / jnp.maximum(u_norm, 1e-10)
    return new_p / jnp.linalg.norm(new_p) * p_norm - p
```

AdamW has no such property. The concern is that RL under AdamW drifts the parameter norms `GatedNorm` was tuned against, in a way that would not appear on Delphi or Qwen because neither has this optimizer/architecture coupling. This is a hypothesis, not an observed failure — nothing has been run yet.

## What this is not

**Not a memory-efficiency request.** AdamH's state is `ScaleByAdamHState(count, mu, nu)` — two moments, identical to Adam. It is a conditioning change, not a footprint change, and it does not relax the memory floor for a 67B-parameter model under FSDP2 with EP=1.

## Port considerations

Not a mechanical translation:

- The update computes per-parameter norms, so it needs the full parameter tensor. Under FSDP2 each rank holds a shard, so a faithful port has to decide whether the norm is global (requiring a collective per parameter per step) or per-shard (a different optimizer). That choice changes the algorithm, not just the implementation.
- Rank-3+ tensors are handled by `jax.vmap` of the 2-D logic over the leading axis. A 256-expert MoE keeps its expert weights exactly there, so this path is load-bearing rather than incidental.
- `learning_rate` is baked into the transformation rather than applied downstream, which does not match how `optimizer_config` is threaded today.

## Asks

1. Is an AdamH port to MarinSkyRL wanted, and if so is the sharded-norm question settled somewhere already?
2. If not, is the AdamW-on-an-AdamH-trained-model risk considered acceptable, and is it documented anywhere? Happy to record the answer in `docs/grug-fsdp2-training.md` alongside the other Grug constraints.
3. Is `GatedNorm`'s compensation expected to be robust to a norm-drifting optimizer, or is it specifically load-bearing for AdamH's invariant?

Context: marin-community/marin#7786 tracks the Snowball RL campaign, and marin-community/MarinSkyRL#110 established the Grug FSDP2 path and its constraint list.

@yonromai flagging you as the author of #110 — happy to be told this is a non-issue.

**Comment by yonromai** [comment]
Thanks for flagging, Ben!

> @yonromai flagging you as the author of https://github.com/marin-community/MarinSkyRL/pull/110 — happy to be told this is a non-issue.

Cool, no I think this is an issue indeed. Working on a fix.
